raft/src/test/java/org/apache/kafka/raft/MockLog.java

```diff
@@ -177,14 +181,16 @@ public void appendAsFollower(Records records) {
 
     @Override
     public Records read(long startOffset, OptionalLong maxOffsetOpt) {
-        List<LogBatch> batches = readBatches(startOffset, maxOffsetOpt);
-        if (batches.isEmpty())
-            return MemoryRecords.EMPTY;
-
         ByteBuffer buffer = ByteBuffer.allocate(512);
-        for (LogBatch batch : batches) {
-            buffer = batch.writeTo(buffer);
+
+        long maxOffset = maxOffsetOpt.orElse(endOffset());
```

Comment:
I fixed this func to make partial batches be returned as well.

Reply:
Hmm... Can you clarify the intent? It is true we may return an incomplete batch with `KafkaMetadataLog`, but it would just get discarded. We would never receive only a part of the batch.

Reply:
I agree, in practice all quorum-fetch's start offset would be at the batch boundary.

But in unit test we do not necessarily mock fetch request that way, and then `readCommitted -> log.read` is used to update the high-watermark and hence need to make sure that hwm is updated to the middle of a batch when that happens too.

I felt it is more flexible that way to cover the case if we ever get quorum-fetch not at batch boundary as well. But if we think this would never be the case (that fetch records always start at batch boundary --- btw today we do not strictly check against that case in handle-fetch I think) we can modify the unit test back to be only getting quorum-fetch starting at batch boundary. LMK

Reply:
Discussed offline with @guozhangwang and we decide to unblock and decide whether we want to revert this change later.